[FYL2003] iP

FYL2003 opened the pull request:
### A Test Heading for GFMD

* A test on bullet list for *
+ A test on bullet list for +
- A test on bullet list for -
- A test on ~~cancel things in list~~

1. number list first item
2. number list second item

```java
public class Test {
    public static void main(String[] args) {
        System.out.println("This is used for Java testing");
    }
}

```

- [x] Finished CS2103 Level-9
- [ ] Level-10 branching
- [ ] Check style

Emoji testing:
🥉 🗡️ 

> This is a testing blockquote.
>
> > Testing nested quote

[Testing link to Google](https://www.google.com.sg/?client=safari&channel=ipad_bm)


`Some random inline code`

__bold test__
~~strikethrough test~~
_abc_

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

FYL2003 committed:
add .class file in gitignore

FYL2003 committed:
rename the name of chatting bot

FYL2003 committed:
complete level 0 requirements

FYL2003 committed:
complete level 1 requirements

FYL2003 committed:
implementing list feature

FYL2003 committed:
add command list detection

FYL2003 committed:
add Task class and basic feature

FYL2003 committed:
change string array to class and testing if it works

FYL2003 committed:
fix index problem

FYL2003 committed:
finish implementing Task class in current stage

FYL2003 committed:
changed indentation

FYL2003 committed:
finish editing todo class

FYL2003 committed:
finish editing deadline class

FYL2003 committed:
finish editing event class, modify deadline class

FYL2003 committed:
add test cases according to level 4 sample output

FYL2003 committed:
change array to arraylist

FYL2003 committed:
modified logic of decideTask factory method

FYL2003 committed:
add exception about empty description after todo, deadline and event

FYL2003 committed:
add exception about garbage input

FYL2003 committed:
modified test case

FYL2003 committed:
move all task mangement to task class

FYL2003 committed:
modified test case, add delete method

FYL2003 committed:
change command (only for bye and list) to case insensitive

FYL2003 committed:
test on reading files

FYL2003 committed:
add logic for creating storage file if it is not found

FYL2003 committed:
create new check logic when the data file not exist

FYL2003 committed:
add file loading function for todo and deadline event

FYL2003 committed:
finish load file

FYL2003 committed:
finish load functionality, rewrite after each iteration

FYL2003 committed:
Merge branch 'branch-Level-7'

FYL2003 committed:
add DateTime to deadline only

FYL2003 committed:
Merge branch 'branch-Level-8'

FYL2003 committed:
create OOP files in need

FYL2003 committed:
Add basic UI

FYL2003 committed:
Implement load method in Storage class

FYL2003 committed:
complete storage save method

FYL2003 committed:
complete bye and list command

FYL2003 committed:
fix bye keyword that never terminate

FYL2003 committed:
add keyword mark and unmark

FYL2003 committed:
add keyword todo, make inner class public

FYL2003 committed:
add keyword event, fix formatting issue

FYL2003 committed:
solve output string issue for keyword mark and unmark
add keyword deadline

FYL2003 committed:
add delete keyword

FYL2003 committed:
handle random input using exception

FYL2003 committed:
handled empty description exception,
delete useless testing print statement

FYL2003 committed:
replace the old file with more oop version

FYL2003 committed:
modify logic of database reader

FYL2003 committed:
change testing input file

FYL2003 committed:
Merge branch 'branch-MoreOOP'

FYL2003 committed:
add package

FYL2003 committed:
Merge branch 'branch-Packages'

FYL2003 committed:
update database

FYL2003 committed:
Merge branch 'add-gradle-support'

FYL2003 committed:
add JUnit test for parser and ui

FYL2003 committed:
change gradle settings to make jar files

FYL2003 committed:
add JavaDoc for half the class

FYL2003 committed:
Merge branch 'branch-A-JavaDoc'

FYL2003 committed:
check with coding CodingStandard

FYL2003 committed:
Merge branch 'branch-A-CodingStandard'

FYL2003 committed:
add parser for keyword find
start ui for keyword find

FYL2003 committed:
finish find keyword

FYL2003 committed:
Merge branch 'branch-Level-9'

TechRaj commented on src/main/java/greetbot/EmptyDescriptionException.java:
```diff
@@ -0,0 +1,22 @@
+ */
+public class EmptyDescriptionException extends Exception {
+    
+    final private String MESSAGE;
```
idk if it violates the principles or not, but I think private comes first (?)

TechRaj commented on src/main/java/greetbot/GreetBot.java:
```diff
@@ -0,0 +1,153 @@
+        return ui.showMarked(tasks.get(position), tasks.getLength());
+    }
+
+    private String markAsNotDone(int position) {
```
Good use of camel caps!

TechRaj commented on src/main/java/greetbot/Task.java:
```diff
@@ -0,0 +1,143 @@
+ */
+abstract class Task {
+    protected String description;
+    private final DateTimeFormatter RUNNINGFORMAT = DateTimeFormatter.ofPattern("MMM d yyyy");
```
Good job on using final. The other guy didnt bother lol

TechRaj commented on src/main/java/greetbot/Parser.java:
```diff
@@ -0,0 +1,52 @@
+     * @param input The argument for mark/unmark/delete keyword.
+     * @return The position to mark/unmark/delete.
+     */
+    public static int parseMarkUnmarkDelete(String input) {
```
Seems like a messy name. Maybe can shorten it while maintaining the same functionality?

reginateh replied:
Second this!

TechRaj commented:
Overall good job! Just some nitpicks

FYL2003 committed:
follow until JavaFX step 4

FYL2003 committed:
Link GUI with bot

FYL2003 committed:
Merge branch 'branch-Level-10'

reginateh commented on src/main/java/greetbot/EmptyDescriptionException.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    /**
+     *  Returns a String that shows the command has empty description.
+     * @return Description of the exception.
```
You could add an empty line between description and parameter section.

reginateh commented on src/main/java/greetbot/GreetBot.java:
```diff
@@ -0,0 +1,167 @@
+    public String getResponse(String input) throws RandomInputException, EmptyDescriptionException{
+        String[] segment = Parser.parseCommand(input);
+        String keyword = segment[0];
+        if (keyword.equals("BYE")) {
+            isRunning = false;
+            ui.closeInput();
+            storage.saveData(tasks);
+            return ui.farewellUser();
+        } else if (keyword.equals("LIST")) {
+            return ui.showList(this.tasks);
+        } else if (keyword.equals("MARK")) {
+            return this.markAsDone(Parser.parseMarkUnmarkDelete(segment[1]) - 1);
+        } else if (keyword.equals("UNMARK")) {
+            return this.markAsNotDone(Parser.parseMarkUnmarkDelete(segment[1]) - 1);
+        } else if (keyword.equals("TODO")) {
+            if (segment.length == 1) {
+                throw new EmptyDescriptionException("OOPS!!! The description of todo cannot be empty.");
+            }
+            return this.addTodo(Parser.parseTodo(segment[1]));
```
Would it be better to use `switch` instead of `if-else`?

reginateh commented on src/main/java/greetbot/DialogBox.java:
```diff
@@ -0,0 +1,60 @@
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
```
You ordered your imports nicely!

reginateh reviewed (commented):
Overall very clean code!

FYL2003 committed:
add checkstyle file
modify file violating the style

FYL2003 committed:
Merge branch 'branch-CheckStyle'

FYL2003 committed:
Add assertions in the code base

Add assertion at 2 place in Storage.java file to
check if the file exists after try to create the
database at load() function and before try to save
data into the database at saveData() function.

As a result, we are sure that the file always
exists when we access it.

FYL2003 committed:
Merge pull request #1 from FYL2003/branch-A-Assertions

Add assertions in the code base

FYL2003 committed:
Merge branch 'master' into branch-A-CodeQuality

FYL2003 committed:
add javadoc and remove magic number

Add javadoc for constructor in GreetBot.java and
EmptyDescriptionException.java. Remove magic number
in GreetBot.java.

Magic number is replaced by final field variable
to imporve readability.

FYL2003 committed:
remove magic number

Remove magic number in Parser.java, replace it to
variable names which is self explainary.

FYL2003 committed:
add javadoc and remove more magic numbers

add javadoc on constructor of RandomInputException.java
remove all the magic number in Storage.java to make
it more self explanatory.

FYL2003 committed:
remove magic number

remove all magic numbers in Task.java to make the
code self explainatory.

Improve readability of the code.

FYL2003 committed:
Merge pull request #2 from FYL2003/branch-A-CodeQuality

Branch a code quality

FYL2003 committed:
Add javadoc and add keyword shortcut

Add javadoc for method in Ui.java. Add shortcut
for the following keywords:list, mark, unmark, todo,
event, deadline, delete.

The user need to type the full command keyword before,
after this update, they are able to use shortcut and
it is easier to remember.

FYL2003 committed:
Merge pull request #3 from FYL2003/branch-C-FriendlierSyntax

Add javadoc and add keyword shortcut

FYL2003 committed:
Add javadoc for some methods

Add javadoc for showDelete() and showFind() in Ui.java,
and findTasks() and showTasks() in TaskList.java.

The descriptions added will better explain how the
program works.

FYL2003 committed:
Remove magic number, change code to meet SLAP

Remove all possible magic number of strings and change
it to self explanatory names. Fix SLAP issue in GreetBot.java
by adding method checkEmptyDescription for handling
empty desceription exceptions.

This allows better readability and make the program
less bug prone.

FYL2003 committed:
Add title for GUI

Add the title to make sure the product name is on
the title bar of the GUI.

FYL2003 committed:
Merge branch 'branch-FixSLAP'

FYL2003 committed:
Add GUI screenshot

FYL2003 committed:
Remove unnecessary store, fix problem for event keyword

Remove the unnecessary store during initialzing the
greetbot in GreetBot.java. Fix problem of accidentally
cut off description string for from part and to part.

This commit fix a major bug in event keyword.

FYL2003 committed:
Add user guide

Modified user guide and add repective picture for
showing the sample.

FYL2003 committed:
Handle more exception

Add exception handling for cases where the number
format is wrong or tring to manipulate position out
of the task list.

This prevent the GUI to sliently treat those command
and will return some error message for the user to
know why their command is wrong.

FYL2003 committed:
Handle more error

Add logic for handle wrong number format, trying
to modify element out of bound or missing argument
for task.

This modification will tell the user more error message
instead of respond nothing.

FYL2003 committed:
Add logic for saving data

Add logic to save data everytime when the tasklist
changes. This will prevent loss of data when the programme
accidentally shut down.

FYL2003 committed:
Modified README

Modified README file and add explanation for exit
command.

FYL2003 committed:
Change README

Correct the name from "duke.jar" to "greetbot.jar"

FYL2003 committed:
Modify README

Add sample command for keyword "mark" for the users
to see how the bot marks task.

FYL2003 committed:
Modify README

Add sample command for keyword "unmark" for the users
to see how the bot unmarks task.

FYL2003 committed:
Modify README

Add sample command for keyword "todo" for the users
to see how the bot adds todo task.

FYL2003 committed:
Modify README

Add sample command for keyword "event" for the users
to see how the bot adds event task.

FYL2003 committed:
Modify README

Add sample command for keyword "deadline" for the users
to see how the bot adds deadline task.
